segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -511,6 +516,43 @@ public void mergeSegments(MergeSegments mergeSegments) {
                     });
     }
 
+    @Override
+    public void mergeSegmentsBatch(WireCommands.MergeSegmentsBatch mergeSegments) {
+        final String operation = "mergeSegmentsBatch";
+
+        List<String> sources = mergeSegments.getSources();
+        for (String s : sources) {
+            if (!verifyToken(s, mergeSegments.getRequestId(), mergeSegments.getDelegationToken(), operation)) {
+                return;
+            }
+        }
+
+        log.info(mergeSegments.getRequestId(), "Merging Segments batch {} ", mergeSegments);
+
+        Function<String, CompletableFuture<Long>> mergeFunction = source ->
+                Futures.handleCompose(segmentStore.mergeStreamSegment(mergeSegments.getTarget(),
+                        source, TIMEOUT), (r, e) -> {
+                    if (e != null) {
+                        if (Exceptions.unwrap(e) instanceof StreamSegmentMergedException) {
+                            log.info(mergeSegments.getRequestId(), "Stream segment is already merged '{}'.",
+                                    sources);
+                            return segmentStore.getStreamSegmentInfo(mergeSegments.getTarget(), TIMEOUT)
+                                    .thenApply(SegmentProperties::getLength);
```

Comment:
I don't believe this is returning what you think it does. What do you do with this return value on the Controller side?

In line 546, you are returning the length of the Target Segment immediately after merging the transaction. That value is guaranteed to be correct even if there had been additional modifications to the segment after the call.

In this line (539-540), you are invoking `getInfo` on the Target Segment and get the current length. This is by no means guaranteed to be the length of the Target Segment immediately after the merge. It is the current length of the Segment. Since this Txn was merged, it's quite possible that other modifications happened to the Target Segment (new appends, Txns, etc) so this value is semantically different than the one you get back at 546.